src/PointwiseFunctions/Hydro/EquationsOfState/Spectral.cpp

```diff
@@ -0,0 +1,317 @@
+    result[i] /= (1.0 + i);
+  }
+  return result;
+}
+}  // namespace
+
+namespace EquationsOfState {
+Spectral::Spectral(const double reference_density,
+                   const double reference_pressure,
+                   std::vector<double> coefficients, const double upper_density)
+    : reference_density_(reference_density),
+      reference_pressure_(reference_pressure),
+      integral_coefficients_(compute_integral_coefficients(coefficients)),
+      gamma_coefficients_(std::move(coefficients)),
+      x_max_(log(upper_density / reference_density)),
+      gamma_of_x_max_(gamma(x_max_)),
+      integral_of_gamma_of_x_max_(integral_of_gamma(x_max_)),
+      number_of_quadrature_coefs_(3) {
+  // Setup table of specific energies
+  // Use (coefsx2) quadrature points
+  quadrature_weights_ = {0.3607615730481386, 0.4679139345726910,
+                         0.1713244923791704};
+  quadrature_points_ = {0.6612093864662645, 0.2386191860831969,
+                        0.9324695142031521};
+  const auto n_points_epsilon = static_cast<size_t>(ceil(2.0 * x_max_) + 1.0);
+  const double delta_x = x_max_ / (n_points_epsilon - 1.0);
+  table_of_specific_energies_.resize(n_points_epsilon);
+  table_of_specific_energies_[0] =
+      reference_pressure_ / reference_density_ / (gamma_coefficients_[0] - 1.0);
+  for (size_t i = 1; i < n_points_epsilon; i++) {
+    table_of_specific_energies_[i] = table_of_specific_energies_[i - 1];
+    // Use Gaussian quadrature to calculate the specific internal
+    // energy at the next point, separated from the old point
+    // by delta_x in log(density)
+    // From the 1st law of thermodynamics,
+    // d(epsilon)/dx = (Pressure)/(reference_density)*exp(-x)
+    const double x0 = (i - 1.) * delta_x;
+    for (size_t q = 0; q < number_of_quadrature_coefs_; q++) {
+      const double xp = x0 + (1.0 + quadrature_points_[q]) * (delta_x / 2.0);
+      const double xm = x0 + (1.0 - quadrature_points_[q]) * (delta_x / 2.0);
+      table_of_specific_energies_[i] +=
+          quadrature_weights_[q] * delta_x / (2.0 * reference_density) *
+          (exp(-xp) * pressure_from_log_density(xp) +
+           exp(-xm) * pressure_from_log_density(xm));
+    }
+  }
+}
+
+EQUATION_OF_STATE_MEMBER_DEFINITIONS(, Spectral, double, 1)
+EQUATION_OF_STATE_MEMBER_DEFINITIONS(, Spectral, DataVector, 1)
+
+Spectral::Spectral(CkMigrateMessage* /*unused*/) {}
+
+void Spectral::pup(PUP::er& p) {
+  EquationOfState<true, 1>::pup(p);
+  p | reference_density_;
+  p | reference_pressure_;
+  p | integral_coefficients_;
+  p | gamma_coefficients_;
+  p | x_max_;
+  p | gamma_of_x_max_;
+  p | integral_of_gamma_of_x_max_;
+  p | number_of_quadrature_coefs_;
+  p | quadrature_weights_;
+  p | quadrature_points_;
+  p | table_of_specific_energies_;
+}
+
+// this evaluates the power series
+// Int_0^x Gamma(xx) dxx = Sum_{n=0}^N \frac{gamma_n}{n+1} xx^{n+1}
+double Spectral::integral_of_gamma(const double x) const {
```

Comment:
I think this is the same as `evaluate_polynomial` from `Utilities/Math.hpp` (except for an overall factor of `x`).